cmd/operator-sdk/bundle/create.go

```diff
@@ -15,116 +15,320 @@
+			"If --output-dir is also set, the original files will not be overwritten")
+	fs.StringVarP(&c.imageBuilder, "image-builder", "b", "docker",
 		"Tool to build container images. One of: [docker, podman, buildah]")
-	cmd.Flags().StringVarP(&c.defaultChannel, "default-channel", "e", "",
+	fs.StringVarP(&c.defaultChannel, "default-channel", "e", "",
 		"The default channel for the bundle image")
+}
 
-	return cmd
+func (c *bundleCreateCmd) setDefaults() (err error) {
+	projectName := filepath.Base(projutil.MustGetwd())
+	if c.directory == "" {
+		c.directory = filepath.Join(catalog.OLMCatalogDir, projectName)
+		// Avoid discrepancy between packageName and directory if either is set
+		// by only assuming the operator dir is the packageName if directory isn't set.
+		c.packageName = projectName
+	}
+	// Clean and make paths relative for less verbose error messages.
+	if c.directory, err = filepath.Abs(c.directory); err != nil {
+		return err
+	}
+	// Set outputDir in any case so we make the operator-registry file generator
+	// write 'manifests/' every time, and handle cleanup logic in runBuild().
+	if c.outputDir == "" {
+		c.outputDir = filepath.Dir(c.directory)
+	}
+	if c.outputDir, err = filepath.Abs(c.outputDir); err != nil {
+		return err
+	}
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	if c.directory, err = filepath.Rel(wd, c.directory); err != nil {
+		return err
+	}
+	if c.outputDir, err = filepath.Rel(wd, c.outputDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c bundleCreateCmd) validate(args []string) error {
+	if c.directory == "" {
+		return fmt.Errorf("--directory must be set")
+	}
+	if c.packageName == "" {
+		return fmt.Errorf("--package must be set")
+	}
+	if c.generateOnly {
+		if len(args) != 0 {
+			return errors.New("the command does not accept any arguments if --generate-only=true")
+		}
+	} else {
+		if len(args) != 1 {
+			return errors.New("a bundle image tag is a required argument if --generate-only=true")
+		}
+	}
+	return nil
+}
+
+// runGenerate generates a bundle.Dockerfile, and manifests/ and metadata/ dirs,
+// always overwriting their contents.
+func (c bundleCreateCmd) runGenerate() error {
+	err := bundle.GenerateFunc(c.directory, c.outputDir, c.packageName, c.channels, c.defaultChannel, true)
+	if err != nil {
+		return fmt.Errorf("error generating bundle image files: %v", err)
+	}
+	return nil
+}
+
+// runBuild runs the equivalent of runGenerate then builds a bundle image. If
+// a manifest/, metadata/ or bundle.DockerFile do not exist, they are removed.
+func (c bundleCreateCmd) runBuild() error {
+	rootDir := filepath.Dir(c.directory)
+	metadataDir := filepath.Join(rootDir, bundle.MetadataDir)
+	manifestsDir := filepath.Join(rootDir, bundle.ManifestsDir)
+
+	// Clean up transient files once the image is built, as they are no longer
+	// needed.
+	if !isExist(manifestsDir) {
+		defer makeRmFunc(manifestsDir)()
```

Comment:
Is there a reason we need to return a function and call it, rather than just defering a function directly? e.g.

`defer remove(manifestsDir)`

Reply:
Nope, just a forgotten remnant of refactoring.